Fixed #27807 -- Made username validation configurable through AUTH_USERNAME_VALIDATORS setting

saykharng opened the pull request:
[ticket-27807](https://code.djangoproject.com/ticket/27807)

Solution allows Username Validation to be configured through AUTH_USERNAME_VALIDATORS setting.
The setting follows similar approach to password validators with slight difference. Username Validator must be callable like core validators and requires a help_text method (returns a string explaining the requirement of validation).

- Added USERNAME_VALIDATORS config in settings.
- Username Validator should be callable, like core validators.
- Username Validator should have help_text() method, which is to display the requirement to the user.
- Added help_text method on UnicodeUsernameValidator and ASCIIUsernameValidator
- Added help_text method on UnicodeUsernameValidator and ASCIIUsernameValidator
- Added Username Validation page in docs/topics
- Added test in auth_tests

jacobtylerwalls commented:
Hi Shekhar, another (prospective) first time Django contributor here. Thought I would help out a fellow first timer by commenting on your documentation.

jacobtylerwalls commented on docs/topics/auth/customizing.txt:
```diff
@@ -444,6 +444,32 @@ a :class:`~django.db.models.ForeignKey` or
+
+.. currentmodule:: django.contrib.auth
+
+Django username validator can be customize using the ``settings.AUTH_USERNAME_VALIDATORS``. You can create validators.py and write your own username validator class. For example::
```
can be customized

jacobtylerwalls commented on docs/topics/auth/customizing.txt:
```diff
@@ -444,6 +444,32 @@ a :class:`~django.db.models.ForeignKey` or
+        Username validator must subclass from :class:`~django.core.validators.RegexValidator` class.
+            
+
+Then, in ``settings.py`` file you can add ``AUTH_USERNAME_VALIDATORS`` and intantiate an object of CustomUsernameValidator. For example::
```
The rest of this file wraps lines after 79 characters, so make sure you do that too!

jacobtylerwalls commented on docs/topics/auth/customizing.txt:
```diff
@@ -444,6 +444,32 @@ a :class:`~django.db.models.ForeignKey` or
+    
+    class CustomUsernameValidator(RegexValidator):
+        regex = r'^[\w.@+-']+\Z'
+        message = ("This value may contain only Enlight letters, numbers, and @/./+/-/'/_ charecters.")
```
English letters

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +260,29 @@ def test_ascii_validator(self):
+    regex = r"^[\w.@'+-]+\Z"
+    message = _(
+                'Enter a valid username. This may contain only English letters,'
+                "numbers, and @/./+/-/_/' charecters."
```
characters

jacobtylerwalls commented on docs/ref/settings.txt:
```diff
@@ -3007,6 +3007,19 @@ The list of validators that are used to check the strength of user's passwords.
+
+Default: ``[]`` (Empty list)
+
+The list of validators that are used to check the restricted charecter in username.
```
character

jacobtylerwalls commented on django/contrib/auth/validators.py:
```diff
@@ -23,3 +24,48 @@ class UnicodeUsernameValidator(validators.RegexValidator):
+    def __init__(self, validator_type):
+        self.validators = []
+        self.validator_type = validator_type
+    
```
flake8 is telling you you have whitespace in an empty line here. To view all the times flake8 found this, go to your pull request, view the automatic checks at the bottom, select Details next to the flake8 test, click Console, scroll bottom where you will see file names and line numbers and the specific problem.

jacobtylerwalls commented on django/conf/global_settings.py:
```diff
@@ -521,6 +521,8 @@ def gettext_noop(s):
 
 AUTH_PASSWORD_VALIDATORS = []
 
+AUTH_USERNAME_VALIDATORS = []
```
Now, to the functionality, my first thought about this approach is that users will expect this to behave like `AUTH_PASSWORD_VALIDATORS`, which is a list containing a dict for each validator, as here:
https://docs.djangoproject.com/en/3.0/topics/auth/passwords/#enabling-password-validation

Did you consider the value in implementing something similar? You would avoid having to implement a list-like object with __len__ and the like, and avoid asking users to import custom classes in settings.py. For more on that:
https://stackoverflow.com/questions/7382722/initializing-custom-objects-in-django-settings

jacobtylerwalls commented on docs/topics/auth/customizing.txt:
```diff
@@ -444,6 +444,32 @@ a :class:`~django.db.models.ForeignKey` or
+    
+    from myapp.validators import CustomUsernameValidator                
+    
+    AUTH_USERNAME_VALIDATORS = [ CustomUsernameValidator() ]
```
Here is where we shouldn't be asking users to instantiate objects in global_settings.py. Notice no other objects are instantiated there.

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +260,29 @@ def test_ascii_validator(self):
+    @override_settings(AUTH_USERNAME_VALIDATORS=[CustomUsernameValidator()])
+    def test_update_validators(self):
+        username_validators = validators.UsernameValidatorList(RegexValidator)
+        self.assertEqual('CustomUsernameValidator', username_validators[0].__class__.__name__) 
```
more common would be `self.assertIsInstance`

saykharng commented:
> Hi Shekhar, another (prospective) first time Django contributor here. Thought I would help out a fellow first timer by commenting on your documentation.
Hi Jacob, Thank you! I will go through each comment and make changes.

saykharng commented:
> Hi Shekhar, another (prospective) first time Django contributor here. Thought I would help out a fellow first timer by commenting on your documentation.

Hi Jacob, my sincere apologies on taking long time on this. 

Currently, I am following the django core validators pattern where the validator should be callable. Therefore implementing  validators with __call__ method. Is this ok? or should I make it exactly like a password validators with validate() function and make necessary changes?

Please let me know.

jacobtylerwalls commented:
Hi Shekhar!
Since this is a new feature, I think the next step is for you to post on django-developers to find consensus on the approach. See more here: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/
Cheers, Jacob

saykharng commented:
> Hi Shekhar!
> Since this is a new feature, I think the next step is for you to post on django-developers to find consensus on the approach. See more here: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/
> Cheers, Jacob

Thank you!

saykharng commented:
> > Hi Shekhar!
> > Since this is a new feature, I think the next step is for you to post on django-developers to find consensus on the approach. See more here: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/
> > Cheers, Jacob
> 
> Thank you!

I have opened a discussion but waiting on feedback.

AntonOfTheWoods commented:
Just thought I'd cheer on this ticket and say I'm eagerly awaiting it getting into django :-).

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+Integrating validation
+----------------------
+
+There are few function in ``django.contrib.auth.username_validation`` that
```
functions

jacobtylerwalls commented on django/contrib/auth/validators.py:
```diff
@@ -23,3 +30,38 @@ class UnicodeUsernameValidator(validators.RegexValidator):
+    def help_text(self):
+        return ngettext(
+            "Your username must contain at least %(min_length)d character.",
+            "Your username must contain at least %(min_length)d characters.",
```
maybe you were basing double quotes off of existing code, but going forward django style is single quotes where possible :-)

jacobtylerwalls commented on django/contrib/auth/username_validation.py:
```diff
@@ -0,0 +1,50 @@
+        try:
+            klass = import_string(validator['NAME'])
+        except ImportError:
+            msg = "The module in NAME could not be imported: %s. Check your AUTH_USERNAME_VALIDATORS setting."
```
single quotes

jacobtylerwalls commented on tests/=:
```diff
@@ -0,0 +1,3 @@
+Collecting isort
```
I don't understand why this file was checked in

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,73 @@ def test_ascii_validator(self):
+                v(valid)
+        for invalid in invalid_usernames:
+            with self.subTest(invalid=invalid):
+                with self.assertRaises(ValidationError):
```
prefer `with self.assertRaisesMessage(ValidationError, msg):` according to this checklist:

https://docs.djangoproject.com/en/3.0/internals/contributing/writing-code/coding-style/

jacobtylerwalls commented:
Cool stuff @saykharng, I expect this needs release notes in docs/releases/3.2.txt, since this really is a new feature even though the ticket was originally classified as a bug (because of what the documentation promised at the time). Make sure to unset the patch needs improvement flag on the ticket next time you push to make sure you end up in the review queue.

jacobtylerwalls commented on django/conf/global_settings.py:
```diff
@@ -524,6 +524,10 @@ def gettext_noop(s):
 AUTH_PASSWORD_VALIDATORS = []
 
+AUTH_USERNAME_VALIDATORS = [
+    {'NAME': 'django.contrib.auth.validators.UnicodeUsernameValidator'},
```
the indentation and line breaks here don't match the documentation you provided

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+    Validates whether the username contains ASCII letters, in addition
+    to @,.,+,-, and _.
+
+.. class:: UsernameMinimumLengthValidator(min_length=3)
```
your class defaults min_length=0, not 3

jacobtylerwalls replied:
versionadded tag

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+=============================
+
+.. module:: django.contrib.auth.username_validation
+
```
.. versionadded:: 3.2

https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+you can call from your own forms or other code to integrate username
+validation.
+
+.. function:: username_validators_help_texts(username_validators=None)
```
versionadded tag here and following

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+.. class:: UnicodeUsernameValidators
+    
+    Validates whether the username contains Unicode characters, in addition
+    to @,.,+,-, and _. This is the default validator.
```
the docs in django.docs.ref.contrib.auth.txt enclose each special character in double grave accents like ``@`` so that they are distinguishable from the commas and jump off the page -- I suggest doing that here, too.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+  anything other than the English letters, numbers and @/./+/-/_.
+* ``UsernameMinimumLengthValidator``, which checks whether the username meets a minimum
+  length. This validator is configured with a custom option it now requires
+  the minimum length to be nine characters, instead of the default eight.
```
the minimum length to be three characters, instead of the default zero.

which sort of raises the question whether the default should be something other than zero!

jacobtylerwalls replied:
Recent changes look good! This one still says nine characters instead of the default eight when I think it should be: three characters instead of the default one.

saykharng replied:
I have fixed this issue and committed code. Forgot to mention here.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,130 @@
+
+Any items in the ``OPTIONS`` in :setting:``AUTH_USERNAME_VALIDATORS`` for your
+validator will be passed to the constructor. All constructor arguments should
+have default value.
```
have a default value

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,73 @@ def test_ascii_validator(self):
+        self.assertEqual(help_texts[1], length)
+
+
+@override_settings(AUTH_USERNAME_VALIDATORS=[
```
Why is this override applied to the entire class when you then override it with the simpler, default setting in line 311? It seems to me you could have no override here, then override only for `test_customized_validation` and that way you would be more directly checking in `test_get_default_username_validators` that the default was populated correctly. Imagine: what if somebody changes global_settings.py later, would your test for the default validators fail?

saykharng replied:
Thanks a lot. This feedback was an eye opener. I will now think about this every time I write test.

jacobtylerwalls commented:
And to answer your question, I like that you made the validators directly callable, since that was the existing behavior of `UnicodeUsernameValidator`. 👍

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+        
+        def get_help_text(self):
+            return _(
+                "Your username must onctain at least %(min_length)d characters."
```
contain

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+validator will be passed to the constructor. All constructor arguments should
+have a default value.
+
+Here's a  basic example of a validator, with one optional setting::
```
one space before "basic"

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+
+Validation is controlled by the :setting:`AUTH_USERNAME_VALIDATORS` setting.
+The default for the setting is an ``UnicodeUsernameValidator`` which
+validates whether given username contains English letters, numbers and @/./+/-/_
```
"whether the given username"

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+Validation is controlled by the :setting:`AUTH_USERNAME_VALIDATORS` setting.
+The default for the setting is an ``UnicodeUsernameValidator`` which
+validates whether given username contains English letters, numbers and @/./+/-/_
+characters or not.
```
instead of "or not" it would be simpler to just say "whether the given username contains only English letters, numbers, and ``@``,``.``,``+``,``-``,``_`` characters.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+
+This example enables two included validators:
+
+* ``UnicodeUsernameValidator``, which checks whether username contains
```
"whether the username"

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+Integrating validation
+----------------------
+
+There are few functions in ``django.contrib.auth.username_validation`` that
```
either "a few" or "several", not "few" by itself

jacobtylerwalls commented on django/contrib/auth/username_validation.py:
```diff
@@ -0,0 +1,50 @@
+    in an <ul>.
+    """
+    help_texts = username_validators_help_texts(username_validators)
+    help_items = format_html_join('', '<li>{}</li>', ((help_text, ) for help_text in help_texts))
```
no space:
``(help_text,)``

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,71 @@ def test_ascii_validator(self):
+class UsernameValidationTests(SimpleTestCase):
+    def test_get_default_username_validators(self):
+        validators_list = get_default_username_validators()
+        self.assertIsInstance(validators_list[0], validators.UnicodeUsernameValidator)
```
I think you should do more than just test that the function `get_default_username_validators()` gets the validators. Rather, I think the pipeline all the way down to `AbstractUser` needs to be tested, since you made changes there. (What if somebody refactors `AbstractUser` later?) This would involve creating some users.

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,71 @@ def test_ascii_validator(self):
+                'min_length': 3,
+            }},
+        ]
+        ASCII_and_unicode = _(
```
I would ensure this variable name begins with a lowercase letter somehow, like`unicode_help_text`

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+  username is valid, or raise a 
+  :exc:`~django.core.exceptions.ValidationError` with an error message if the
+  username is not valid.
+* ``get_help_text(self)``: provide a help text to explain the requirements to
```
Are you sure `get_help_text(self)` works? The validators you contribute have a method called `help_text(self)` instead. The documentation should be consistent, but more importantly, I suggest writing a test that implements a custom validator so that we would have caught this already.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,140 @@
+        def get_help_text(self):
+            return _(
+                "Your username must onctain at least %(min_length)d characters."
+                "% {'min_length': self.min_length}
```
the % formatting happens outside the `ngettext()` or `_()` in your patch, not inside

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,71 @@ def test_ascii_validator(self):
+    def test_username_validators_help_text(self):
+        validator_config = [
+            {'NAME': 'django.contrib.auth.validators.UnicodeUsernameValidator'},
+            {'NAME': 'django.contrib.auth.validators.UsernameMinimumLengthValidator', 'OPTIONS': {
```
break line before 'OPTIONS'

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -259,3 +264,71 @@ def test_ascii_validator(self):
+
+    @override_settings(AUTH_USERNAME_VALIDATORS=[
+        {'NAME': 'django.contrib.auth.validators.UnicodeUsernameValidator'},
+        {'NAME': 'django.contrib.auth.validators.UsernameMinimumLengthValidator', 'OPTIONS': {
```
break line before 'OPTIONS'

jacobtylerwalls commented on docs/ref/settings.txt:
```diff
@@ -3029,6 +3029,19 @@ The list of validators that are used to check the strength of user's passwords.
+.. setting:: AUTH_USERNAME_VALIDATORS
+
+``AUTH_USERNAME_VALIDATORS``
+----------------------------
```
I think this needs a short help text (like above for password validation) and then a .. versionadded:: 3.2

saykharng commented:
> heya @saykharng I have a few more suggestions on tests to include and some documentation fixes. I'll let a more experienced contributor mark this as ready for checkin once they've given it their own review 👍

Hi Jacob, thanks a lot for your feedback and sorry it's taking a while.

I have learned a lot about development and django by working on this feature. I have added few more test and made all the changes as suggested. Appreciate your time. :)

saykharng commented:
> Happy to do it, @saykharng, I left a few more comments that might say "outdated" because they were made in between your force-pushes, but I gave them a quick scan and I think they're all still relevant.
> 
> Also, be careful about the files you check into the repo, I noticed you checked in a .swp file.

Thank you!

* Fixed the docs based on comments (looking at a html page was much easier to find  issue.)
* Removed username_validator variable
* Removed word 'English' from 'English letters' in UnicodeUsernameValidator

AntonOfTheWoods commented on django/contrib/auth/validators.py:
```diff
@@ -14,6 +15,12 @@ class ASCIIUsernameValidator(validators.RegexValidator):
 
+    def help_text(self):
+        return _(
+            'Enter a valid username. This value may contain only English letters, '
```
From my experience speaking those languages and as a native speaker of English, I also vote for 'letters from the Latin alphabet'. Some languages (like Chinese) tend to use "English" for everything to do with Europe, but none of the European languages I speak do in my experience!

saykharng replied:
Thank you! I have made the changes as suggested.

matthiask commented on django/contrib/auth/validators.py:
```diff
@@ -23,3 +30,38 @@ class UnicodeUsernameValidator(validators.RegexValidator):
+
+
+@deconstructible
+class UsernameMinimumLengthValidator:
```
I'm unsure whether there's a good reason for this validator, since `django.core.validators.MinLengthValidator` (https://docs.djangoproject.com/en/3.1/ref/validators/#minlengthvalidator) exists already. 

I know there's a MinimumLengthValidator in the `password_validation.py` module but maybe that one is a bit redundant too.

saykharng replied:
Thank you for the feedback. Apologies for extreme delay in response.

I have made suggested changes other than moving validators to username_validation.

matthiask commented on docs/ref/settings.txt:
```diff
@@ -3029,6 +3029,26 @@ The list of validators that are used to check the strength of user's passwords.
+
+The list of validators that are used to check username requirement.
+See :ref:`username-validation` for more details. By default, 
+UnicodeUsernameValidator is used to perform the validation.
```
Use ``:class:`~django.contrib.auth.validators.UnicodeUsernameValidator` `` here

saykharng replied:
Fixed

matthiask commented on docs/ref/settings.txt:
```diff
@@ -3029,6 +3029,26 @@ The list of validators that are used to check the strength of user's passwords.
+    
+    [
+        {
+            'NAME': 'contrib.auth.validators.UnicodeUsernameValidator'
```
"django." is missing

saykharng replied:
Fixed

matthiask commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,148 @@
+their behavior.
+
+Validation is controlled by the :setting:`AUTH_USERNAME_VALIDATORS` setting.
+The default for the setting is a ``UnicodeUsernameValidator`` which
```
``":class:`~...`"`` again

saykharng replied:
Fixed

matthiask commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,148 @@
+        {
+            'NAME': 'django.contrib.auth.validators.UsernameMinimumLengthValidator',
+            'OPTIONS':{
+                'min_length':3,
```
Spaces after `:` missing

saykharng replied:
Fixed

matthiask commented on docs/topics/auth/username.txt:
```diff
@@ -0,0 +1,148 @@
+                'Your username must contain at least %(min_length)d 
+                character.',
+                'Your username must contain at least %(min_length)d
+                characters.',
```
Strange line breaks

saykharng replied:
Fixed

matthiask reviewed (commented):
- I really like replacing "English" with "Latin alphabet"
- Maybe you could move the existing username validators to the `username_validation` module too since you have been inspired by the `password_validation` module (and of course leave backwards compatibility shims in `django.contrib.auth.validators`.) *(Not important, just a suggestion.)*

jacobtylerwalls commented on tests/auth_tests/test_validators.py:
```diff
@@ -284,7 +284,8 @@ def test_username_validators_help_text(self):
                 'OPTIONS': {
-                    'min_length': 3,
+                    'limit_value': 3,
+                    'message': 'This password is too short',
```
username, not password. :-)

saykharng replied:
> glad to see this moving along. some minor fixes after your last push.

Thank you! that is so embarrassing. I actually read everything in the docs this time in html.

-- Fixed all the typos and the line break.
-- Added the new method help_text for Unicode and ASCII username validator in the docs
-- Added the UsernameMinimumLength validator in list of Validators as well.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -117,32 +122,18 @@ Any items in the ``OPTIONS`` in :setting:`AUTH_USERNAME_VALIDATORS` for your
-                self.min_length
-            ) % {'min_length': self.min_length}
+                'Your username must contain at least %(limit_value)d
+                character.','Your username must contain at least 
```
I think the other reviewer's comment about the line breaks meant you should get each string on one line, in full. This code as it is wouldn't compile. You don't have to wrap at 79 chars if you're quoting a code sample.

saykharng replied:
fixed.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -70,12 +73,14 @@ Django includes three validators:
-    The minimum length can be customized with the ``min_length`` parameter.
+    Validates whether the username meets a minimum length using 
+    ``limit_value``. This validator is extended from 
+    ``django.core.validators.MinLengthValidator``. The text given in ``message``
```
add ``:class:`~...``

saykharng replied:
fixed.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -117,32 +122,18 @@ Any items in the ``OPTIONS`` in :setting:`AUTH_USERNAME_VALIDATORS` for your
 
-Here's a basic example of a validator, with one optional setting::
+Here's a basic example of a validator. This validator extends 
+``django.core.validators.MinLengthValidator``::
```
add ``:class:`~...``

saykharng replied:
fixed.

jacobtylerwalls commented on docs/topics/auth/username.txt:
```diff
@@ -38,7 +39,8 @@ Username validation is configured in the
             'OPTIONS':{
-                'min_length':3,
+                'input_value': 3,
+                'message': 'Password is too short.'
```
username, not password :-)

saykharng replied:
fixed.

jacobtylerwalls reviewed (commented):
glad to see this moving along. some minor fixes after your last push.

jacobtylerwalls commented:
Hi again. I think to move this forward what you likely need is proof that this has support to get into Django. I was glad to see you post on the mailing list when I suggested it in July, but perhaps the question you asked was so narrow it didn't introduce the subject properly.

I think you need to ask more broadly whether this feature should be fashioned after password validation with a dict of validators in `global_settings.py`, and if so, whether your implementation is on track, and further, whether it should be documented with an entire new page almost identical to the password validation page, or whether they should be integrated. One respondent on the ticket modeled an approach similar to yours, so that's promising.

Minor things to make this easier to review: please update the body of the description of the PR with the ticket number (ticket-27807), and make the PR title describe the approach you used, something like: "Made username validation configurable through AUTH_USERNAME_VALIDATORS setting"

Even if people end up suggesting a different approach, you will have still made a significant contribution in terms of putting forth a concrete approach for folks to look at. 👍

saykharng commented:
> Hi again. I think to move this forward what you likely need is proof that this has support to get into Django. I was glad to see you post on the mailing list when I suggested it in July, but perhaps the question you asked was so narrow it didn't introduce the subject properly.
> 
> I think you need to ask more broadly whether this feature should be fashioned after password validation with a dict of validators in `global_settings.py`, and if so, whether your implementation is on track, and further, whether it should be documented with an entire new page almost identical to the password validation page, or whether they should be integrated. One respondent on the ticket modeled an approach similar to yours, so that's promising.
> 
> Minor things to make this easier to review: please update the body of the description of the PR with the ticket number ([ticket-27807](https://code.djangoproject.com/ticket/27807)), and make the PR title describe the approach you used, something like: "Made username validation configurable through AUTH_USERNAME_VALIDATORS setting"
> 
> Even if people end up suggesting a different approach, you will have still made a significant contribution in terms of putting forth a concrete approach for folks to look at. 👍

Thank you again. I have started a group conversation and updated the detail in PR as well. :)

carltongibson reviewed (commented):
Hi @saykharng. Thanks for the input here, but I think we have to close this as wontfix. 

The suggested fix here adds a setting, as signal, and sub-api to validators — all of which are significant changes, and I think none of which are justified by the issues. 

_How do I customize username validation?_ First and foremost, it's not clear you should: `UnicodeUsernameValidator` is a good default. But if you want to, use a custom user model. If you can't do that, add the validation you want at the form level, or in `clean()`, or if you really must monkey patch the field in a `ready()` handler, and so on. 

There are lots of option that don't need so much structure. 

I hope that makes sense. I will update the ticket accordingly.

saykharng commented:
> Hi @saykharng. Thanks for the input here, but I think we have to close this as wontfix.
> 
> The suggested fix here adds a setting, as signal, and sub-api to validators — all of which are significant changes, and I think none of which are justified by the issues.
> 
> _How do I customize username validation?_ First and foremost, it's not clear you should: `UnicodeUsernameValidator` is a good default. But if you want to, use a custom user model. If you can't do that, add the validation you want at the form level, or in `clean()`, or if you really must monkey patch the field in a `ready()` handler, and so on.
> 
> There are lots of option that don't need so much structure.
> 
> I hope that makes sense. I will update the ticket accordingly.

@carltongibson 
Appreciate your time and thank you for the feedback. 
I did learn a lot about Django, tests and documentation while working on it.

In future, I will open up discussion/share ideas earlier.

Regards
Shekhar
